postprocessing/science/spectral_corner_frequency_from_surface_xdmf.py

```diff
@@ -0,0 +1,384 @@
+    else:
+        p = Pool(nprocs)
+        result = p.map(ReadFaultWithPool, chunks)
+        p.close()
+        p.join()
+        return np.concatenate(result, axis=1).T[::stepsize]
+    
+def ReadSurfaceWithPool(chunk):
+    return surfacexdmf.ReadDataChunk(chunk[2], int(chunk[0]), int(chunk[1]))[timeIndices[0]:timeIndices[1]]
+
+def ReadFaultWithPool(chunk):
+    return faultxdmf.ReadDataChunk(chunk[2], int(chunk[0]), int(chunk[1]))[timeIndicesFault[0]:timeIndicesFault[1]]
+
+def ComputeTriangleMidpoints(geom, connect):
+    """Generates an array with coordinates of triangle midpoints (same dimension as connect)"""
+    xyz = np.zeros_like(connect, dtype=float)
+    xyz = (1./3.)*(geom[connect[:,0],:]+geom[connect[:,1],:]+geom[connect[:,2],:])   
+    return xyz
+
+def CalculateSlipCentroid(faultxdmf, events=0):
+    if events == 0:
+        ASl = faultxdmf.ReadData("ASl", idt=lastTimeStep).T
+    elif events == 1:
+        ASl = faultxdmf.ReadData("ASl", idt=int(lastTimeStep/2)).T   
+    else:
+        ASl = faultxdmf.ReadData("ASl", idt=lastTimeStep).T - faultxdmf.ReadData("ASl", idt=int(lastTimeStep/2)).T 
+    return np.average(faultxyz, axis=0, weights=ASl)
+
+def ComputeBackazimuth(xyz, centroid):
+    ba = np.arctan2(xyz[:,0]-centroid[0], xyz[:,1]-centroid[1]) % (2*np.pi)
+    ba = np.where(ba < np.pi, ba+np.pi, ba-np.pi)
+    return ba
+
+def RotateToRadialTransversal(u, v, xyz, centroid):
+    ba = ComputeBackazimuth(xyz, centroid)
+    ba = ba.reshape(-1,1)
+    r = -u * np.sin(ba) - v * np.cos(ba)
+    t = -u * np.cos(ba) + v * np.sin(ba)
+    return r, t
+
+def ComputeCornerFrequencyParallel(component, pphases, windowduration):
+    chunk_inds = np.int_(np.linspace(0, component.shape[0], nprocs+1))
+    chunks = []
+    for i in range(nprocs):
+        chunks.append([component[chunk_inds[i]:chunk_inds[i+1]], pphases[chunk_inds[i]:chunk_inds[i+1]], 
+                     windowduration[chunk_inds[i]:chunk_inds[i+1]]])
+    chunks = [tuple(i) for i in chunks]  
+       
+    p = Pool(nprocs)
+    result = p.map(ComputeCornerFrequencyArray, chunks)
+    p.close()
+    p.join()
+    return np.concatenate(result)        
+
+def ComputeCornerFrequencyArray(tuples):
+    # Needed for parallel computing of corner frequencies
+    fc_arr = np.zeros(tuples[0].shape[0])
+    for i in range(0, fc_arr.size):
+        fc_arr[i] = ComputeSingleCornerFrequency(tuples[0][i], dt, pphase=tuples[1][i], 
+                                                 windowduration=tuples[2][i], endfreq=args.maxFreq[0])
+    return fc_arr
+
+def ComputeSingleCornerFrequency(velocity, dt, pphase=0, windowduration=0, endfreq=1.0):
+    """Preproccesing of a single waveform before calling fc_Qzero_gridsearch to determine the corner frequency"""
+
+    # preprocessing (taper and padding)
+    if windowduration != 0:
+        data = cut_window_and_pad(velocity, windowduration, dt, pphase=pphase, alpha=0.05)
+    else:
+        data = velocity
```

Comment:
I would also tapper the data in this case.

Reply:
Done